thriftbp/client_middlewares.go

```diff
@@ -313,3 +318,75 @@ var (
 	_ thrift.ClientMiddleware = SetDeadlineBudget
 	_ thrift.ClientMiddleware = BaseplateErrorWrapper
 )
+
+// PrometheusClientMiddleware returns middleware to track Prometheus metrics
+// specific to the Thrift client.
+//
+// It emits the following prometheus metrics:
+//
+// * thrift_client_active_requests gauge with labels:
+//
+//   - thrift_service: the serviceSlug arg
+//   - thrift_method: the method of the endpoint called
+//   - thrift_slug: the service being contacted
+//
+// * thrift_client_latency_seconds histogram with labels above plus:
+//
+//   - thrift_success: "true" if err == nil, "false" otherwise
+//
+// * thrift_client_requests_total counter with all labels above plus:
+//
+//   - thrift_exception_type: the human-readable exception type, e.g.
+//     baseplate.Error, etc
+//   - thrift_baseplate_status: the numeric status code from a baseplate.Error
+//     as a string if present (e.g. 404), or the empty string
+//   - thrift_baseplate_status_code: the human-readable status code, e.g.
+//     NOT_FOUND, or the empty string
+func PrometheusClientMiddleware(serviceSlug, serverSlug string) thrift.ClientMiddleware {
+	return func(next thrift.TClient) thrift.TClient {
+		return thrift.WrappedTClient{
+			Wrapped: func(ctx context.Context, method string, args, result thrift.TStruct) (_ thrift.ResponseMeta, err error) {
+				start := time.Now()
+				activeRequestLabels := prometheus.Labels{
+					serviceLabel: serviceSlug,
+					methodLabel:  method,
+					slugLabel:    serverSlug,
+				}
+				clientActiveRequests.With(activeRequestLabels).Inc()
+
+				defer func() {
+					var exceptionTypeLabel, baseplateStatusCode, baseplateStatus string
+					successLbl := strconv.FormatBool(err == nil)
+					if err != nil {
+						exceptionTypeLabel = strings.TrimPrefix(fmt.Sprintf("%T", err), "*")
+
+						var bpErr baseplateError
+						if errors.As(err, &bpErr) {
+							code := bpErr.GetCode()
+							baseplateStatusCode = strconv.FormatInt(int64(code), 10)
+							baseplateStatus := baseplate.ErrorCode(code).String()
+							if baseplateStatus == "<UNSET>" {
+								baseplateStatus = ""
+							}
+						}
+					}
+
+					labels := prometheus.Labels{
+						serviceLabel: serviceSlug,
+						methodLabel:  method,
+						successLabel: successLbl,
+						slugLabel:    serverSlug,
+					}
+					clientLatencyDistribution.With(labels).Observe(time.Since(start).Seconds())
+					labels[baseplateStatusCodeLabel] = baseplateStatusCode
+					labels[exceptionLabel] = exceptionTypeLabel
+					labels[baseplateStatusLabel] = baseplateStatus
```

Comment:
🔕 I don't entirely like that the labels above get a literal and these don't... It increases the "cognitive burden" of the code because you need to track the mutations in your head to figure out the keys that are set and whether anything is overridden.

There's also a chance that the strings have duplicate values -- the literal prevents this bug at compile time, but mutating the map would not.

I would probably just make two maps (names are just for example):

```
histogramLabels := prometheus.Labels{...}
detailLabels := prometheus.Labels{...}
```